Scripts/Playthrough.cs

```diff
@@ -374,6 +341,235 @@ private bool IsConnected()
+                OnStopTyping?.Invoke(message);
+            });
+
+            _room.OnMessage<Events.ProblemEvent>("problem", (message) =>
+            {
+                Logger.LogWarning($"Received `problem` event: {JsonConvert.SerializeObject(message)}");
+            });
+
+            _room.OnMessage<string>("pong", (message) =>
+            {
+                _pingCount = 0;
+                OnPingSuccess?.Invoke();
+                SetConnectionState(ConnectionState.Connected);
+            });
+
+            _room.OnLeave += (code) =>
+            {
+                Logger.Log($"Connection closed. Attempting to reconnect to Playthrough.");
+                MainThreadConsumer.Instance?.StartCoroutine(TryToReconnect());
+            };
+        }
+
+        private void OnReady()
+        {
+            SetConnectionState(ConnectionState.Connected);
+
+            _onReadyCallback?.Invoke();
+
+            _calledByDisconnect = false;
+            _reconnectionTryCount = 0;
+
+            Task.Run(FirePing);
+        }
+
+        private IEnumerator TryToReconnect()
+        {
+            if (_calledByDisconnect)
+            {
+                SetConnectionState(ConnectionState.Disconnected);
+                yield break;
+            }
+
+            // cannot allow multiple reconnection coroutines to run concurrently
+            if (IsReconnecting())
+            {
+                yield break;
+            }
+
+            Logger.Log("Trying to reconnect.");
+            SetConnectionState(ConnectionState.Reconnecting);
+
+            MainThreadConsumer.Instance.StartCoroutine(Reconnect());
+        }
+
+        private bool IsReconnecting()
+        {
+            return _connectionState == ConnectionState.Reconnecting;
+        }
+
+        private IEnumerator Reconnect()
+        {
+            bool roomExpired = false;
+            ColyseusRoom<object> room = null;
+
+            while (room == null)
+            {
+                _reconnectionTryCount++;
+
+                if (_reconnectionTryCount <= MAXIMUM_RECONNECTION_ATTEMPTS)
+                {
+                    // Need to set the protocol to secure every time.
+                    // Colyseus resets to non-secure after even one failed attempt
+                    _client.Settings.useSecureProtocol = true;
+
+                    Task<ColyseusRoom<object>> reconnectionTask;
+                    if (roomExpired)
+                    {
+                        // If the room has expired, attempt to create a new one.
+                        reconnectionTask = _client.JoinOrCreate("chat", _roomOptions);
+                    }
+                    else
+                    {
+                        // Try just reconnecting otherwise.
+                        reconnectionTask = _client.Reconnect(_room.Id, _room.SessionId);
+                    }
+
+                    yield return new WaitUntil(() => reconnectionTask.IsCompleted);
+
+                    try
```

Comment:
I'm not sure if this `try/catch` block does anything as there's no code within the `try` block that could throw an error

I'm not so good at coroutines/async but instead I think we want something like:

```c#
if (reconnectionTask.Status != TaskStatus.Faulted) {
  // happy case
} else {
  // sad case
  // we can get the exception from the task (note it seems to be wrapped in `AggregateException`)
  reconnectionTask.Exception;
}
```

ref: https://learn.microsoft.com/en-us/dotnet/standard/parallel-programming/exception-handling-task-parallel-library

Reply:
Thanks, that definitely works better! I can confirm the try/catch was working - accessing the task.Result while in a try still counts as an exception and it would trigger catch in my breakpoint tests. But this is a lot more readable and sensible